// @mgcrea/fastify-session/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CRYPTO_SPLIT_CHAR = exports.ErrorWithCode = exports.createError = exports.SessionStore = exports.MemoryStore = exports.Session = exports.SessionCrypto = void 0;
const fastify_plugin_1 = __importDefault(require("fastify-plugin"));
const plugin_1 = require("./plugin");
var crypto_1 = require("./crypto");
Object.defineProperty(exports, "SessionCrypto", { enumerable: true, get: function () { return crypto_1.SessionCrypto; } });
var session_1 = require("./session");
Object.defineProperty(exports, "Session", { enumerable: true, get: function () { return session_1.Session; } });
var store_1 = require("./store");
Object.defineProperty(exports, "MemoryStore", { enumerable: true, get: function () { return store_1.MemoryStore; } });
Object.defineProperty(exports, "SessionStore", { enumerable: true, get: function () { return store_1.SessionStore; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "createError", { enumerable: true, get: function () { return utils_1.createError; } });
Object.defineProperty(exports, "ErrorWithCode", { enumerable: true, get: function () { return utils_1.ErrorWithCode; } });
Object.defineProperty(exports, "CRYPTO_SPLIT_CHAR", { enumerable: true, get: function () { return utils_1.CRYPTO_SPLIT_CHAR; } });
exports.default = (0, fastify_plugin_1.default)(plugin_1.plugin, {
    fastify: '3.x',
    name: 'fastify-session',
});
